src/borda_clj/client.clj

```diff
@@ -0,0 +1,61 @@
+(ns borda-clj.client
+  (:use [slingshot.slingshot :only [throw+ try+]])
+  (:require [aleph.http :as http]
+            [byte-streams :as bs]
+            [cheshire.core :as json]))
+
+(defn collect
+  "Adds the given measurement (values and dimensions) to the given hashmap and caps the size of the hashmap at the given max-buffer-size"
+  [measurements max-buffer-size dimensions values]
+  (let [key (into (sorted-map) dimensions)]
+    (if (contains? measurements key)
+      ; TODO: support stuff other than SUM here (e.g. AVG, MIN, MAX, etc.)
+      (update measurements key (partial merge-with +) values) ; merge with existing in buffer
+      (if (< (count measurements) max-buffer-size)
+        (assoc measurements key values) ; space available, add to buffer
+        (do (print "borda buffer full, discarding measurement" dimensions values)
+            measurements))))) ; buffer full
+
+(defn reducing-submitter
+  "Returns two functions. The first is a reducing submitter that collects and
+   aggregates measurements and sends them using the given send function at the
+   specified interval (in milliseconds) and limits the size of pending
+   measurements to max-buffer-size. The second is a function that flushes the
+   buffer and stops the background thread that does the submitting."
+  [max-buffer-size interval send on-send-error]
+  (let [measurements  (atom {})
+        running       (atom true)
+        submit        (fn [dimensions values] (swap! measurements collect max-buffer-size dimensions values))
+        resubmit      (fn [m] (doseq [[dimensions values] m] (submit dimensions values)))
+        flush         (fn [on-flush-error]
+                        (let [[m] (reset-vals! measurements {})]
+                          (if (not-empty m)
+                            (try
+                              (send m)
+                              (catch Throwable e
+                                (on-flush-error m e))))))
+        stop          (fn [] (reset! running false) (flush on-send-error))]
+        ; periodically send to borda
+    (future (while @running (do
+                              (Thread/sleep interval)
+                              (flush (fn [m e] (on-send-error m e) (resubmit m))))))
+    [submit stop]))
+
+(defn http-sender
+  "Returns a function that can be used as the 'send' parameter to
+   reducing-submitter and that sends the measurements to the specified stream at
+   the given URL via HTTP"
+  [stream url]
+  (fn [measurements socket-timeout conn-timeout]
```

Comment:
[`send`](https://github.com/getlantern/borda-clj/pull/1/files#diff-00f5ab632e4080f183618ccb8ef5a023R34) is not passing a `socket-timeout` nor a `conn-timeout`.  Since these parameters seem quite specific to `http-sender`, and since I don't suppose being able to set different timeouts for different `send` calls is a requirement, I think they should be moved to the outer argument list, together with `stream` and `url`.

Reply:
Good point, will fix. While documenting the timeouts, I realized that these aren't the right timeouts to use, so I changed them to connection-timeout and request-timeout.